### PR TITLE
New model for resource usage optimisation in RCPSP problem

### PR DESCRIPTION
**Pull request by g-poveda**
-  new model for resource usage optimisation
-  include a new parameters to params CP (optimisation_level) passed to minizinc solver in the generic minizinc solve function.

### CODE REVIEW

**Review by nhuet — CHANGES_REQUESTED** [review]
I do not have the skills to comment the new minizinc model.
Except for the change to make to the `copy` method, nothing mandatory.
Still an example (or better tests) would be nice to see in action these changes. Currently these new functions are used nowhere.

File: discrete_optimization/rcpsp/minizinc/rcpsp_single_mode_mzn.mzn [hunk 1]
```diff
@@ -167,11 +167,8 @@ solve
    minimize objective+sec_objective;
 
 %-----------------------------------------------------------------------------%
-output [
-    "Start times = ", show(s), "\n",
-    "makespan = ", show(objective), "\n",
-    "sec objective = ", show(sec_objective), "\n",
-    "rr=", show(rr)];
+output ["makespan = ", show(objective), "\n",
+        "sec objective = ", show(sec_objective), "\n"];
```
[the remark this answers is not in the document]
Comment (g-poveda) [reply discrete_optimization/rcpsp/minizinc/rcpsp_single_mode_mzn.mzn:172]: changing log output (less verbose, notably for big rcpsp instance.

File: discrete_optimization/generic_tools/cp_tools.py [hunk 2]
```diff
@@ -60,6 +60,7 @@ class ParametersCP:
     free_search: bool
     multiprocess: bool
     nb_process: int
+    optimisation_level: int
```
[the remark this answers is not in the document]
Comment (g-poveda) [reply discrete_optimization/generic_tools/cp_tools.py:63]: done in 2a3dbd8f886c72accec4049a6bf6001942af10ef

File: discrete_optimization/rcpsp/minizinc/rcpsp_multi_mode_resource_feasibility_mzn.mzn [hunk 3]
```diff
@@ -223,17 +217,19 @@ predicate post_noc_cumu(int: i, int: j) =
...
+                            forall(i,j in Act where i < j /\ can_act_be_disjunct(i, j))(
+                                if are_act_disjunct(i, j) then
+                                    % Activities are disjunct all mode combinations
+                                    post_noc_disj(i, j)
+                                else
+                                    % Activities might be disjunct
+                                    post_noc_rres(i, j)
+                                endif
+                            ))
+           endif;
```
[the remark this answers is not in the document]
Comment (g-poveda) [reply discrete_optimization/rcpsp/minizinc/rcpsp_multi_mode_resource_feasibility_mzn.mzn:235]: parameter non hard coded anymore : https://github.com/airbus/discrete-optimization/pull/45/commits/cd6980756249cef30d70c7bf1837a4d0f8ecc88f

File: discrete_optimization/generic_tools/cp_tools.py [hunk 4]
```diff
@@ -219,6 +224,7 @@ def solve(
                 if parameters_cp.multiprocess
                 else None,
                 free_search=parameters_cp.free_search,
+                optimisation_level=parameters_cp.optimisation_level,
             )
         logger.info("Solving finished")
         logger.debug(result.status)
```
[the remark this answers is not in the document]
Comment (g-poveda) [reply discrete_optimization/generic_tools/cp_tools.py:231]: testing all available values in https://github.com/airbus/discrete-optimization/pull/45/commits/ff4b0bfe36414ef476546774c32320301f5e0b4f

**Comment by nhuet** [comment]
LGTM once the workflow succeeds. (see my minor remark about minizinc model name [here](https://github.com/airbus/discrete-optimization/pull/45#discussion_r992336347).)

**Comment by g-poveda** [comment]
Ok, tests finally passed with more timeout in solvers.

**Pull request merged**